src/devices/timer-src.js

```diff
@@ -0,0 +1,199 @@
+  // the precise time, but the repeated calls will reliably be triggered at
+  // consistent intervals.
+  //
+  // The timer can also create Repeater objects that allow holders to schedule
+  // events at regular intervals even though individual callbacks can be
+  // arbitrarily delayed. The Repeaters have access to their startTime and
+  // interval as well as the latest time we were polled. This allows them to
+  // reschedule.
+  function buildRecurringTimeout(startTime, interval) {
+    let disabled = false;
+    const r = harden({
+      schedule(callback) {
+        if (disabled) {
+          return;
+        }
+        const lastPolled = getLastPolled();
+        // nextTime is the smallest startTime + N * interval after lastPolled
+        const nextTime =
+          lastPolled + interval - ((lastPolled - startTime) % interval);
+        deadlines.add(nextTime, callback, r);
+      },
+      disable() {
+        disabled = true;
+      },
+    });
+    return r;
+  }
+
+  return buildRecurringTimeout;
+}
+
+function setup(syscall, state, helpers, endowments) {
+  function makeRootDevice({ SO, getDeviceState, setDeviceState }) {
+    const initialDeviceState = getDeviceState();
+
+    // A MultiMap from times to schedule objects, with repeaters when present
+    // { time: [{callback}, {callback, repeater}, ... ], ... }
+    const deadlines = initialDeviceState.deadlines || buildTimerMap();
+
+    // The latest time poll() was called. This might be a block height or it
+    // might be a time from Date.now(). The current time is not reflected back
+    // to the user.
+    let lastPolled = initialDeviceState.lastPolled || 0;
+
+    function updateState(time) {
+      insist(
+        time > lastPolled,
+        `Time is monotonic. ${time} must be greater than ${lastPolled}`,
+      );
+      lastPolled = time;
+      setDeviceState(harden({ lastPolled, deadlines }));
+    }
+
+    const innerPoll = curryPollFn(SO, deadlines);
+    const poll = t => {
+      updateState(t);
+      return innerPoll(t);
+    };
+    endowments.registerDevicePollFunction(poll);
+
+    const buildRepeater = curryRepeaterBuilder(deadlines, () => lastPolled);
+
+    // Verify that an alleged callback is a function and has a wake() method.
+    function assertCallbackHasWake(callback) {
+      if (!('wake' in callback)) {
+        throw new TypeError(
+          `callback.wake() does not exist, has ${Object.getOwnPropertyNames(
+            callback,
+          )}`,
+        );
+      }
+      if (!(callback.wake instanceof Function)) {
+        throw new TypeError(
+          `callback[wake] is not a function, typeof is ${typeof callback.wake}, callback has ${Object.getOwnPropertyNames(
+            callback.wake,
+          )}`,
+        );
+      }
+    }
+
+    // The Root Device Node
+    return harden({
+      setWakeup(delaySecs, callback) {
+        assertCallbackHasWake(callback);
```

Comment:
This is going to fail in a real Vat environment (which probably means we need at least a simple integration test). I'd recommend renaming `callback` to `handler`, maybe it's just me but I think of a variable named `callback` as being an invokable function (so `callback instanceof Function === true`).

But in our environment it must be a Presence, which is a placeholder object, whose only special feature is that it lives in a WeakMap which maps it to an `o-NN` object reference that can be used in a syscall.

(I use this pattern in `mailbox-src.js`, in the `registerInboundHandler` function).

The reason this will fail outside of a mock environment is that Presences do not have any properties: you can wrap them in `SO().method(args)` to send a message, but that just sends the method name to the far side. We cannot locally determine whether it has a `.wake` method or not. Someday, maybe, we'll have some cool typechecking to make this better, but for now it's all asynchronous duck-typing.

But I think the way this invokes the "callback"/"handler" is correct, as is the way the handler gets stored. Removing the `asssertCallbackHasWake` is the only bit that needs changing.